stdlib/public/runtime/Metadata.cpp

```diff
@@ -1307,34 +1292,34 @@ class TupleCacheEntry
     return 0; // No single meaningful value
   }
 
-  int compareWithKey(const Key &key) const {
-    // Order by the cheaper comparisons first:
+  friend llvm::hash_code hash_value(const TupleCacheEntry &value) {
+    llvm::SmallVector<const Metadata *, 16> elements;
+    elements.reserve(value.Data.NumElements);
+    for (size_t i = 0, e = value.Data.NumElements; i != e; ++i) {
+      elements.push_back(value.Data.getElement(i).Type);
+    }
+
+    Key key{value.Data.NumElements, elements.data(), value.Data.Labels};
+    return hash_value(key);
```

Comment:
Surely there's a way to do this without having to copy the elements into a local array.  You could use a `TransformRange` to project out the `Metadata*` for `hash_value()`.

Reply:
I couldn't find a nice way to do it, but I didn't know we had `TransformRange`. That looks nice and simple, I'll do that.

Note that entry hashes are only used when resizing the table, so it's not a performance impact for lookups. But it's still definitely worth doing better.